### PR TITLE
Feature/prefix handling retry

### PR DESCRIPTION
**Pull request by shabbyrobe**
This PR adds support for Prefix/Delimiter matching. It makes gofakes3 respond as you'd expect when using the AWS CLI to list folders.

I thought I was finished last night, but I ran into some regressions around creation date and bucket listing when I decided to test a bit further, so I put some work into the test suite. What I have here is a little messy; I intend to clean it up over time as new bits emerge. I like writing high-density test cases with a lot of custom assertions and wrappers that encapsulate the noise. As the API for testing gets more polished over time, the test cases start to read like a specification of behaviour rather than an imperative mechanism for actually running the test. The direction I think the testing API should iterate towards should eventually allow us to use most of the test cases to certify any backend as functioning. This will take some time though.

Most of how I've been verifying my own changes work involves crappy bash scripts with the AWS CLI, which has been a really good way to flush out quirks that break my expectations and to sniff out the expected response, but isn't good for catching regressions. I've attempted to start working this in to a test suite of its own (awscli_test.go), which has some advantages and some drawbacks. On the plus side, it means that we can guarantee various facets of the command line will meet user's expectations, and we are also implicitly testing gofakes3 with botocore this way. On the minus side, it's slow as a wet week, so I'd be reluctant to go crazy adding too many cases that use the CLI.

It might be worth us pulling in an assertion library, too. The little TT struct I added is a cut-down version of my simple assertion lib, but it doesn't report error lines properly. The entire library is small enough to copy and paste around, weighing in at around 150 lines of actual code. This is my preferred approach, but the testify/assert library is another very good option that merits consideration. I would prefer not to use the whole of testify, only the testify/assert bit.

Apologies for the test failures, I have no idea why CircleCI fails the tests while they pass just fine on my machine after a fresh clone.

### CODE REVIEW

**Comment by shabbyrobe** [comment]
Ah I think I found it. Sorry if you've been bombarded with CircleCI emails!

File: backend/s3bolt/backend.go [hunk 1]
```diff
@@ -111,14 +194,36 @@ func (db *Backend) CreateBucket(name string) error {
 func (db *Backend) DeleteBucket(name string) error {
 	return db.bolt.Update(func(tx *bolt.Tx) error {
 		nameBts := []byte(name)
-		b := tx.Bucket(nameBts)
-		if b == nil {
-			return gofakes3.ErrNoSuchBucket
+
+		if bytes.Equal(nameBts, db.metaBucketName) {
+			panic("gofakes3: attempted to delete metadata bucket")
```
[the remark this answers is not in the document]
Comment (shabbyrobe) [reply backend/s3bolt/backend.go]: Hm I think you might be right, it's user input isn't it? panic shouldn't be triggered from user input. I'll fix it, good catch!

File: awscli_test.go [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+func (tc *testCLI) lsFiles(bucket string, prefix string) (items cliLsItems) {
+	tc.Helper()
+
+	prefix = strings.TrimLeft(prefix, "/")
+	out := tc.combinedOutput("s3", "ls", fmt.Sprintf("s3://%s/%s", bucket, prefix))
+
+	scn := bufio.NewScanner(bytes.NewReader(out))
+	for scn.Scan() {
+		cur := scn.Text()
+		dir := cliLsDirMatcher.FindStringSubmatch(cur)
```
[the remark this answers is not in the document]
Comment (shabbyrobe) [reply awscli_test.go]: I don't have a plan to use it in other functions as such, but it's in the global scope so it doesn't have to recompile on every call; it just compiles once on startup. I'm not fussed either way, would you prefer it encapsulated?

[the remark this answers is not in the document]
Comment (johannesboyne) [reply awscli_test.go]: Right, didn't think about the recompile step! Let's keep it there.

**Pull request merged**